source/dataman/CLS/CLSDbUpgrade1Pt3.cpp

```diff
@@ -0,0 +1,134 @@
+#include "CLSDbUpgrade1Pt3.h"
+
+#include <QStringBuilder>
+
+#include "util/AMErrorMonitor.h"
+
+CLSDbUpgrade1Pt3::CLSDbUpgrade1Pt3(const QString &databaseNameToUpgrade, QObject *parent)
+	: AMDbUpgrade(databaseNameToUpgrade, parent)
+{
+}
+
+CLSDbUpgrade1Pt3::~CLSDbUpgrade1Pt3()
+{
+
+}
+
+QStringList CLSDbUpgrade1Pt3::upgradeFromTags() const
+{
+	return QStringList();
+}
+
+bool CLSDbUpgrade1Pt3::upgradeNecessary() const
+{
+	// run update when:
+	//
+	//   - there is no CLSUserConfiguration_table and no CLSUserConfiguration_table_regionsOfInterest
+
+	bool tableExists = databaseToUpgrade_->tableExists("CLSUserConfiguration_table");
+
+	return !tableExists;
+}
+
+bool CLSDbUpgrade1Pt3::upgradeImplementation()
+{
+	// Prologue.
+	////////////////////////////////////////////////
+
+	databaseToUpgrade_->startTransaction();
+
+	// Stage 1:  create table: CLSUserConfiguration_table and CLSUserConfiguration_table_regionsOfInterest
+	///////////////////////////////////////////////////
+
+	QString clsConfigurationTableName = "CLSUserConfiguration_table";
+	QString clsUserConfigurationClassInheritence = "CLSUserConfiguration;AMDbObject;QObject";
+
+	if ( !databaseToUpgrade_->tableExists(clsConfigurationTableName) ) {
+		QStringList columnNames = QStringList() << "AMDbObjectType" << "thumbnailCount" << "thumbnailFirstId" << "name";
+		QStringList columnTypes = QStringList() << "TEXT" << "INTEGER" << "INTEGER" << "TEXT";
+
+		if (!AMDbUpgradeSupport::addTable(databaseToUpgrade_, clsConfigurationTableName, columnNames, columnTypes, true, "CLS User Configuration", clsUserConfigurationClassInheritence)) {
+			databaseToUpgrade_->rollbackTransaction();
+			AMErrorMon::alert(this, CLSDBUPGRADE1PT3_COULD_NOT_CREATE_TABLE, QString("Could not create %1.").arg(clsConfigurationTableName));
+			return false;
+		}
+
+		QString configurationROITableName = QString("%1_regionsOfInterest").arg(clsConfigurationTableName);
+		QStringList ROITableColumnNames = QStringList() << "id1" << "table1" << "id2" << "table2";
+		QStringList ROITableColumnTypes = QStringList() << "INTEGER" << "TEXT" << "INTEGER" << "TEXT";
+
+		if (!databaseToUpgrade_->ensureTable(configurationROITableName, ROITableColumnNames, ROITableColumnTypes)) {
```

Comment:
Oh yes.  That's my bad.  Actually, now that I think about it, would addTable actually be a bad idea since it auto-adds things to the other tables?  You might have been more correct with your original code.